code/dal.py

```diff
@@ -0,0 +1,117 @@
+
+
+def load_from_db(num_emails=100, fetch_all=False):
+    """loads the requested number of emails from the database and dumps the file as csv. For this function to
+    run the mysql database must be set up locally.
+    if fetch_all is false, file is stored as <num_emails>_emails.csv
+    if the fetch_all is true, file is stored as all_emails.csv
+    :param num_emails: The number of emails to be fetched from the db
+    :param fetch_all: Will fetch all the emails from db. Large join query that may take some time"""
+    db_conn = MySQLdb.connect(host='localhost', user='root', passwd='mudit1990', db='EnronAHS')
+    file_path = '../data/'
+    if fetch_all:
+        res = __db_query_all(db_conn)
+        file_name = 'all_emails.csv'
+    else:
+        file_name = str(num_emails) + '_emails.csv'
+        res = __db_query_partial(db_conn, num_emails)
+    data = __clean_data(res)
+    data = data[:,1:] # remove the first column that contains the random id
+    np.savetxt(file_path+file_name, data, fmt='%s', delimiter=',')
+
+
+def __get_appr_filename(num_emails, file_path):
+    """
+    gets the appropriate csv file based on the num_emails required. Tries to select the file with minimum rows that
+    meets the requirement
+    :param num_emails:
+    :param file_path:
+    :return:
+    """
+    files = [f for f in os.listdir(file_path) if os.path.isfile(os.path.join(file_path, f))]
+    max_rows = -1
+    rec_rows = -1
+    for f in files:
+        f_list = f.split('_')
+        try:
+            rows = int(f_list[0])
+            if rows > max_rows:
+                max_rows = rows
+            if rows >= num_emails and (rows < rec_rows or rec_rows == -1):
+                rec_rows = rows
+        except ValueError:
+            continue
+    if rec_rows == -1 and max_rows == -1:
+        return None
+    if rec_rows == -1:
+        rec_rows = max_rows
+    return str(rec_rows) + '_emails.csv'
+
+
+def __clean_data(r):
+    """
+    The function assumes an input as a tuple of tuples, cleans the data and returns a numpy array
+    """
+    data = np.asarray(r)
+    # replace ',' separator in receivers with '|'
+    data[:,2] = np.core.defchararray.replace(data[:, 2], ',', '|')
+    # replace ',' separator in email body with space
+    data[:, 3] = np.core.defchararray.replace(data[:, 3], ',', ' ')
+    # replace '\n' with ' '
+    data[:, 3] = np.core.defchararray.replace(data[:, 3], '\n', ' ')
+    # replace '\t' with ' '
+    data[:, 3] = np.core.defchararray.replace(data[:, 3], '\t', ' ')
+    return data
+
+
+def __db_query_all(db_conn):
+    """
+    :param db_conn:
+    :return: tuples of tuples containing data
+    """
+    query = "select message.mid, sender, group_concat(rvalue) as receivers, body from" + \
+            "EnronAHS.message inner join EnronAHS.recipientinfo on message.mid=recipientinfo.mid group by message.mid;"
+    cur = db_conn.cursor()
+    cur.execute("SET SESSION group_concat_max_len = 100000;")
```

Comment:
Can you explain what is happening here?

Reply:
so basically i'm concatenating the receivers. max limit for group_concat is 1024. That was not enough so I set it higher. I'll put in an explicit comment